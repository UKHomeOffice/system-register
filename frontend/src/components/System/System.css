/*.systemDetails > ul{*/
/*    list-style: none;*/
/*    padding-left: 0;*/
/*}*/

/*.systemDetails{*/
/*    line-height: 1.5;*/
/*}*/

.riskContainer {
  display: flex;
  flex-flow: row wrap;
  justify-content: space-between;
}

.riskSummary {
  background-color: #FAFAFA;
  padding: 20px;
  width: 400px;
  word-wrap: break-word;
}

.riskSummary > h4 {
  margin-top: 0;
  margin-bottom: 16px;

}

.riskSummary > p {
  margin-top: 16px;
  margin-bottom: 10px;
}

.contentBlock {
    color: var(--black);
    font-family: Helvetica, sans-serif;
    border-bottom: solid 1px rgb(191, 193, 195);
}
.contentBlock > h2 {
    font-size: 24px;
    margin-top: 24px;
    margin-bottom: 0;
}

.contentBlock > h1 {
    font-size: 36px;
    margin-bottom: 0;
    margin-top: 32px;
}
.contentBlock > p {
    font-size: 16px;
    margin-top: 8px;
}