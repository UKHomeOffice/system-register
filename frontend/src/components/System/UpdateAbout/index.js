import UpdateAbout from "./UpdateAbout";

export default UpdateAbout;
