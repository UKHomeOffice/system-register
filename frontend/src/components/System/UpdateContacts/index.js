import UpdateContacts from "./UpdateContacts";

export default UpdateContacts;
