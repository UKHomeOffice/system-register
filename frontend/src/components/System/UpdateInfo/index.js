import UpdateInfo from "./UpdateInfo";

export default UpdateInfo;