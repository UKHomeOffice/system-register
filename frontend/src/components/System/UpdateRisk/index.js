import UpdateRisk from "./UpdateRisk";

export default UpdateRisk;
