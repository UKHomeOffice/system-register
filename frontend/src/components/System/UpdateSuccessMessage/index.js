import UpdateSuccessMessage from "./UpdateSuccessMessage";

export default UpdateSuccessMessage;
