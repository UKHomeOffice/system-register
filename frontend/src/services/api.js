import axios from 'axios'
import config from '../config/config'

const api = {
  getAllSystems,
  getSystem,
  updateProductOwner,
  updateCriticality
};

const nullIfEmpty = (value) => value !== "" ? value : null;

async function getAllSystems() {
  const response = await axios.get(`${config.api.url}/systems`);
  const sortedSystems = response.data.systems.sort((a, b) => {
    const x = a.name.toUpperCase();
    const y = b.name.toUpperCase();
    if (x > y) {
      return 1
    } else if (x < y) {
      return -1
    }
    return 0
  })
  return { ...response.data, systems: sortedSystems };
}

async function getSystem(id) {
  const register = await getAllSystems();
  return register.systems.find(s => s.id.toString() === id);
}

async function updateProductOwner(id, data) {
  const response = await axios.post(
    `${config.api.url}/systems/${id}/update-product-owner`,
    { product_owner: nullIfEmpty(data.productOwner) },
    {
      headers: {
        Authorization: `Bearer ${localStorage.getItem("bearer-token")}`,
      }
    });
  return response.data;
}

async function updateCriticality(id, data) {
  const response = await axios.post(
    `${config.api.url}/systems/${id}/update-criticality`,
    { criticality: nullIfEmpty(data.criticality) },
    {
      headers: {
        Authorization: `Bearer ${localStorage.getItem("bearer-token")}`,
      }
    });
  return response.data;
}

export default api;
