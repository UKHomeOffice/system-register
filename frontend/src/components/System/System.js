import React, { useCallback, useEffect, useState } from "react";
import PropTypes from "prop-types";
import { Route, Switch, useHistory, useRouteMatch } from "react-router-dom";
import { filter, isEmpty, map } from "lodash-es";

import PageNotFoundError from "../../components/Errors/PageNotFoundError";
import SecureRoute from "../SecureRoute";
import SystemView from "./SystemView/SystemView";
import UpdateAbout from "./UpdateAbout/UpdateAbout";
import UpdateContacts from "./UpdateContacts";
import UpdateInfo from "./UpdateInfo";
import api from "../../services/api";
import useAsyncError from "../../utilities/useAsyncError";

import "./System.css";

const CHANGE_STATUS_NONE = null;
const CHANGE_STATUS_SUCCESS = "success";

const actionsByField = {
  name: api.updateSystemName,
  description: api.updateSystemDescription,
  aliases: api.updateSystemAliases,
  criticality: api.updateCriticality,
  investmentState: api.updateInvestmentState,
  portfolio: api.updatePortfolio,
  developedBy: api.updateDevelopedBy,
  supportedBy: api.updateSupportedBy,
  businessOwner: api.updateBusinessOwner,
  productOwner: api.updateProductOwner,
  technicalOwner: api.updateTechnicalOwner,
  serviceOwner: api.updateServiceOwner,
  informationAssetOwner: api.updateInformationAssetOwner,
};

const findMatchingActions = (data, fields) =>
  map(
    filter(fields, (field) => field in data),
    (field) => actionsByField[field]
  );

function useReturnToSystemView() {
  const history = useHistory();
  const { url } = useRouteMatch();

  return useCallback(() => {
    history.push(url);
  }, [history, url]);
}

function useUpdateCallbackFactory(id, onChange, setSystem, setChangeStatus) {
  const returnToSystemView = useReturnToSystemView();

  return useCallback(
    (...fields) => async (data) => {
      const actions = findMatchingActions(data, fields);
      for (const action of actions) {
        setSystem(await action(id, data));
      }
      if (!isEmpty(actions)) {
        onChange();
      }
      setChangeStatus(CHANGE_STATUS_SUCCESS);
      returnToSystemView();
    },
    [id, returnToSystemView, onChange, setSystem, setChangeStatus]
  );
}

function System({ portfolios, onChange, onBeforeNameChange }) {
  const {
    path,
    params: { id },
  } = useRouteMatch();
  const [system, setSystem] = useState(null);
  const [changeStatus, setChangeStatus] = useState(CHANGE_STATUS_NONE);
  const throwError = useAsyncError();

  useEffect(() => {
    const fetchData = async () => setSystem(await api.getSystem(id));
    fetchData().catch((e) => {
      throwError(e);
    });
  }, [id, throwError]);

  const createUpdateCallback = useUpdateCallbackFactory(
    id,
    onChange,
    setSystem,
    setChangeStatus
  );
  const handleUpdateInfo = createUpdateCallback(
    "name",
    "description",
    "aliases"
  );
  const handleUpdateAbout = createUpdateCallback(
    "portfolio",
    "criticality",
    "investmentState",
    "developedBy",
    "supportedBy"
  );
  const handleUpdateContacts = createUpdateCallback(
    "businessOwner",
    "technicalOwner",
    "serviceOwner",
    "productOwner",
    "informationAssetOwner"
  );
  const handleCancel = useReturnToSystemView();
  const handleDismiss = useCallback(() => {
    setChangeStatus(CHANGE_STATUS_NONE);
  }, []);

  return (
    <Switch>
      <Route path={`${path}`} exact>
        <SystemView
          system={system}
          status={changeStatus}
          onClose={handleDismiss}
        />
      </Route>
      <SecureRoute path={`${path}/update-info`}>
        <UpdateInfo
          system={system}
          onSubmit={handleUpdateInfo}
          onCancel={handleCancel}
          onBeforeNameChange={onBeforeNameChange}
        />
      </SecureRoute>
      <SecureRoute path={`${path}/update-about`}>
        <UpdateAbout
          system={system}
          portfolios={portfolios}
          onSubmit={handleUpdateAbout}
          onCancel={handleCancel}
        />
      </SecureRoute>
      <SecureRoute path={`${path}/update-contacts`}>
        <UpdateContacts
          system={system}
          onSubmit={handleUpdateContacts}
          onCancel={handleCancel}
        />
      </SecureRoute>
      <Route path="/*" component={PageNotFoundError} />
    </Switch>
  );
}

System.propTypes = {
  portfolios: PropTypes.arrayOf(PropTypes.string).isRequired,
  onChange: PropTypes.func.isRequired,
  onBeforeNameChange: PropTypes.func.isRequired,
};

export default System;
