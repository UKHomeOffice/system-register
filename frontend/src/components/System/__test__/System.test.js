import React from 'react';
import user from "@testing-library/user-event";
import { useKeycloak } from "@react-keycloak/web";
import { render, screen, waitFor } from '@testing-library/react';
import { createMemoryHistory } from "history";
import { Route, Router } from 'react-router-dom';

import System from '../System';
import api from '../../../services/api';

jest.mock('../../../services/api', () => ({
  getSystem: jest.fn(),
  updateProductOwner: jest.fn(),
  updateCriticality: jest.fn(),
}));
jest.mock("@react-keycloak/web", () => ({
  useKeycloak: jest.fn(),
}));

const test_system = {
  name: "Test System",
  criticality: "unknown",
  last_updated: {},
  risks: [],
  aliases: [],
};

describe('<System />', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    api.getSystem.mockResolvedValue(test_system);
  });

  it('renders system view', async () => {
    renderWithHistory("1");

    const element = await screen.findByText('Test System');

    expect(element).toBeInTheDocument();
  });

  describe("when authorized", () => {
    beforeEach(() => {
      useKeycloak.mockReturnValue({
        keycloak: {
          authenticated: true,
        },
        initialized: true,
      });
    });

    it("shows an edit view for info", async () => {
      renderWithHistory("1/update-info");

      const element = await screen.findByText('Test System');

      expect(element).toBeInTheDocument();
    });

    it("shows an edit view for about", async () => {
      renderWithHistory("1/update-about");

      const element = await screen.findByText('Test System');

      expect(element).toBeInTheDocument();
    });

    it("shows an edit view for contacts", async () => {
      renderWithHistory("1/update-contacts");

      const element = await screen.findByText('Test System');

      expect(element).toBeInTheDocument();
    });

    describe("editing contacts", () => {
      it("returns to the system view after a successful update", async () => {
        api.updateProductOwner.mockResolvedValue({ ...test_system, product_owner: "updated owner" });
        const history = createMemoryHistory({
          initialEntries: ["/system/123/update-contacts"],
          initialIndex: 0,
        });
        renderWithHistory(null, { history });
        const productOwnerField = await screen.findByLabelText(/product owner/i);
        const saveButton = screen.getByRole("button", { name: /save/i });

        // noinspection ES6MissingAwait: there is no typing delay
        user.type(productOwnerField, "updated owner");
        user.click(saveButton);

        await waitFor(() => {
          expect(api.updateProductOwner).toBeCalledWith(
            "123",
            expect.objectContaining({
              productOwner: "updated owner",
            })
          );

          expect(history).toHaveProperty("index", 1);
          expect(history).toHaveProperty(
            "location.pathname",
            "/system/123"
          );
        });
      });

      it("returns to system view on cancel", async () => {
        await checkCancelButton("update-contacts");
      });

      it("does not send a request if field values are unchanged", async () => {
        const history = createMemoryHistory({
          initialEntries: ["/system/123/update-contacts"],
          initialIndex: 0,
        });
        renderWithHistory(null, { history });
        await screen.findByText("Test System");
        const saveButton = await screen.findByRole("button", { name: /save/i });

        user.click(saveButton);

        await waitFor(() => {
          expect(history).toHaveProperty("index", 1);
          expect(history).toHaveProperty(
            "location.pathname",
            "/system/123"
          );
        });
        expect(api.updateProductOwner).not.toBeCalled();
      });
    });

    describe("editing system info", () => {
      describe("system name", () => {
        it("returns to system view on cancel", async () => {
          await checkCancelButton("update-info");
        });

        it("returns to the system view after a successful update", async () => {
          api.updateSystemName.mockResolvedValue({ ...test_system, name: "updated system name" });
          const history = createMemoryHistory({
            initialEntries: ["/system/123/update-info"],
            initialIndex: 0,
          });
          renderWithHistory(null, { history });
          const systemNameField = await screen.findByLabelText(/system owner/i);
          const saveButton = screen.getByRole("button", { name: /save/i });

          // noinspection ES6MissingAwait: there is no typing delay
          user.type(systemNameField, "updated system name");
          user.click(saveButton);

          await waitFor(() => {
            expect(api.updateSystemName).toBeCalledWith(
              "123",
              expect.objectContaining({
                name: "updated system name",
              })
            );

            expect(history).toHaveProperty("index", 1);
            expect(history).toHaveProperty(
              "location.pathname",
              "/system/123"
            );
          });
        });
      })
    })

    describe("editing about", () => {
      describe("criticality", () => {
        it("returns to system view on cancel", async () => {
          await checkCancelButton("update-about");
        });

        it("returns to the system view after a successful update", async () => {
          api.updateCriticality.mockResolvedValue({ ...test_system, criticality: "high" });
          const history = createMemoryHistory({
            initialEntries: ["/system/123/update-about"],
            initialIndex: 0,
          });
          renderWithHistory(null, { history });
          const radioButton = await screen.findByLabelText(/high/i);
          const saveButton = screen.getByRole("button", { name: /save/i });

          // noinspection ES6MissingAwait: there is no typing delay
          user.click(radioButton);
          user.click(saveButton);

          await waitFor(() => {
            expect(api.updateCriticality).toBeCalledWith(
              "123",
              expect.objectContaining({
                criticality: "high",
              })
            );

            expect(history).toHaveProperty("index", 1);
            expect(history).toHaveProperty(
              "location.pathname",
              "/system/123"
            );
          });
        });

        it("api is not called if criticality is unchanged", async () => {
          const history = createMemoryHistory({
            initialEntries: ["/system/123/update-about"],
            initialIndex: 0,
          });
          renderWithHistory(null, { history });
          const radioButton = await screen.findByLabelText(/unknown/i);
          const saveButton = screen.getByRole("button", { name: /save/i });

          // noinspection ES6MissingAwait: there is no typing delay
          user.click(radioButton);
          user.click(saveButton);

          await waitFor(() => {
            expect(history).toHaveProperty("index", 1);
            expect(history).toHaveProperty(
              "location.pathname",
              "/system/123"
            );
          });
          expect(api.updateCriticality).not.toBeCalled();
        });
      })
    });
  });
});

async function checkCancelButton(path) {
  const history = createMemoryHistory({
    initialEntries: [`/system/123/${path}`],
    initialIndex: 0,
  });
  renderWithHistory(null, { history });
  const cancelButton = await screen.findByRole("button", { name: /cancel/i });

  user.click(cancelButton);

  expect(history).toHaveProperty("index", 1);
  expect(history).toHaveProperty(
    "location.pathname",
    "/system/123"
  );

  expect(api.updateProductOwner).not.toBeCalled();
  expect(api.updateCriticality).not.toBeCalled(); //todo refactor to loop through all update api methods
}

function renderWithHistory(path, context = {}) {
  const { history } = {
    history: createMemoryHistory({ initialEntries: [`/system/${path}`] }),
    ...context,
  };

  return render(
    <Router history={history}>
      <Route path='/system/:id' component={System} />
    </Router>
  );
}
