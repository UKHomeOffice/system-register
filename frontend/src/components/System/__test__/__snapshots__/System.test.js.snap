// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<System /> matches snapshot with known data 1`] = `
<div
  className="systemDetails centerContent"
>
  <h2>
    System X
  </h2>
  <span
    className="badge criticalityHigh"
  >
    Criticality 
    High
  </span>
  <span
    className="badge investEvergreen"
  >
    Evergreen
  </span>
  <ul>
    <li>
      Portfolio: 
      Portfolio Y
    </li>
    <li>
      <span>
        Last Updated: 
        February 2020
      </span>
    </li>
    <li>
      Aliases: 
      sys X, project Jupiter
    </li>
  </ul>
  <h3>
    Description
  </h3>
  <p>
    The description
  </p>
  <h3>
    Contacts
  </h3>
  <ul>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        System Register Owner
        : 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </span>
    </li>
  </ul>
  <ul>
    <li>
      Developed By: 
      Dev Team
    </li>
    <li>
      Supported By: 
      Support Team
    </li>
  </ul>
  <ul>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Business Owner
        : 
        I own the business
      </span>
    </li>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Technical Owner
        : 
        I own the tech
      </span>
    </li>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Service Owner
        : 
        I own the service
      </span>
    </li>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Product Owner
        : 
        I own the product
      </span>
    </li>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Information Asset Owner
        : 
        I own the data
      </span>
    </li>
  </ul>
  <h3>
    Risk
  </h3>
  <div
    className="riskContainer"
  >
    <div
      className="riskSummary"
    >
      <h4>
        Roadmap
      </h4>
      <span
        className="badge highRiskBackground"
      >
        Risk: 
        High
      </span>
      <p>
        Rationale: 
        there is no plan
      </p>
    </div>
    <div
      className="riskSummary"
    >
      <h4>
        Tech Stack
      </h4>
      <span
        className="badge lowRiskBackground"
      >
        Risk: 
        Low
      </span>
      <p>
        Rationale: 
        Really good tech
      </p>
    </div>
  </div>
</div>
`;

exports[`<System /> matches snapshot with unknown data 1`] = `
<div
  className="systemDetails centerContent"
>
  <h2>
    System X
  </h2>
  <span
    className="badge criticalityUnknown"
  >
    Criticality 
    Unknown
  </span>
  <span
    className="badge investmentUnknown"
  >
    Investment Unknown
  </span>
  <ul>
    <li>
      Portfolio: 
      <strong
        className="highRisk"
      >
        Unknown
      </strong>
    </li>
    <li>
      <span>
        Last Updated: 
        <strong
          className="highRisk"
        >
          Never
        </strong>
      </span>
    </li>
    <li>
      Aliases: 
      None
    </li>
  </ul>
  <h3>
    Description
  </h3>
  <p>
    <strong
      className="highRisk"
    >
      Unknown
    </strong>
  </p>
  <h3>
    Contacts
  </h3>
  <ul>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        System Register Owner
        : 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </span>
    </li>
  </ul>
  <ul>
    <li>
      Developed By: 
      <strong
        className="highRisk"
      >
        Unknown
      </strong>
    </li>
    <li>
      Supported By: 
      <strong
        className="highRisk"
      >
        Unknown
      </strong>
    </li>
  </ul>
  <ul>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Business Owner
        : 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </span>
    </li>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Technical Owner
        : 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </span>
    </li>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Service Owner
        : 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </span>
    </li>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Product Owner
        : 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </span>
    </li>
    <li>
      <span
        className="contact"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        Information Asset Owner
        : 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </span>
    </li>
  </ul>
  <h3>
    Risk
  </h3>
  <div
    className="riskContainer"
  >
    <div
      className="riskSummary"
    >
      <h4>
        Roadmap
      </h4>
      <span
        className="badge unknownRiskBackground"
      >
        Risk: 
        Unknown
      </span>
      <p>
        Rationale: 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </p>
    </div>
    <div
      className="riskSummary"
    >
      <h4>
        Change
      </h4>
      <span
        className="badge unknownRiskBackground"
      >
        Risk: 
        Unknown
      </span>
      <p>
        Rationale: 
        <strong
          className="highRisk"
        >
          Unknown
        </strong>
      </p>
    </div>
  </div>
</div>
`;
